### ISSUE TITLE
Provide ability to customize the timeout for KCP when talking to a workload etcd cluster

### ISSUE THREAD

**Issue by jdef** [kubernetes-sigs/cluster-api#5586]
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

https://github.com/kubernetes-sigs/cluster-api/blob/7e42829e0ccd15d0960ad52cf2a9616cb7c6a72d/controlplane/kubeadm/internal/etcd/etcd.go#L33

Tried to bootstrap a control plane in Singapore, from my local workstation in the US. Ping latency is about 250ms. KCP repeatedly fails to create a functional etcd client. Internal code uses a hardcoded timeout of 2s, w/ no way to override. Bumping this timeout in a forked version of the code resolved the problem.

**What did you expect to happen:**

Expected cluster spinup to be successful over high latency network.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version: 0.3.19 (forked)
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by jdef** [comment]
cc @dims

**Comment by neolit123** [comment]
do you have any logs to prove the problem?

> Tried to bootstrap a control plane in Singapore, from my local workstation in the US. Ping latency is about 250ms. KCP repeatedly fails to create a functional etcd client. Internal code uses a hardcoded timeout of 2s, w/ no way to override. Bumping this timeout in a forked version of the code resolved the problem.

this is the etcd client dial timeout and 2 seconds is also a default in kubeadm:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/etcd/etcd.go#L199

2 seconds should be sufficient, which is proven by your ICMP of 250ms to the host.

https://pkg.go.dev/github.com/coreos/etcd/clientv3
> 	// DialTimeout is the timeout for failing to establish a connection.

more than 2 seconds feels like this local machine should not be used to control the remote machine.

kubeadm has a tracking issue for better control on timeouts, but that work is currently unstaffed and needs design:
https://github.com/kubernetes/kubeadm/issues/2463

**Comment by jdef** [comment]
> do you have any logs to prove the problem?

I have tcpdumps, aforementioned ping logs, and the "unable to create etcd client" message from KCP (after I updated the code to actually log this). I also have the empirical evidence of the problem going away after bumping this timeout. I can't share specific log files, etc. because of our company's OSS policy. But I'm happy to discuss further in this ticket.

> > Tried to bootstrap a control plane in Singapore, from my local workstation in the US. Ping latency is about 250ms. KCP repeatedly fails to create a functional etcd client. Internal code uses a hardcoded timeout of 2s, w/ no way to override. Bumping this timeout in a forked version of the code resolved the problem.
> 
> this is the etcd client dial timeout and 2 seconds is also a default in kubeadm: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/etcd/etcd.go#L199
> 
> 2 seconds should be sufficient, which is proven by your ICMP of 250ms to the host.

No, not proven - sorry. It's not sufficient for high latency WAN. A number of packets are exchanged w/ the apiserver when creating a new etcd client.

> https://pkg.go.dev/github.com/coreos/etcd/clientv3
> 
> > // DialTimeout is the timeout for failing to establish a connection.
> 
> more than 2 seconds feels like this local machine should not be used to control the remote machine.

We don't have the luxury of additional dev machines in every colo to be used as bootstrappers.

> kubeadm has a tracking issue for better control on timeouts, but that work is currently unstaffed and needs design: [kubernetes/kubeadm#2463](https://github.com/kubernetes/kubeadm/issues/2463)

kubeadm, when bootstrapping, is also - generally - run locally (e.g. kubeadm init ..., kubeadm join ...), especially in the context of CAPI. It's probably reasonable for a 2s delay when connecting via localhost. Comparing use case of kubeadm-via-localhost and KCP-via-Internet seems like comparing apples and oranges.

**Comment by timothysc** [comment]
I think it would be possible to have local env/cli-flag override for this.  I can understand the user story, but it's definitely an exception.  

@jdef can you use a cloud region jump box in the interim?

**Comment by fabriziopandini** [comment]
FYI There is prior art for the kubeadm token TTL that can be set via a flag
https://github.com/kubernetes-sigs/cluster-api/blob/7e42829e0ccd15d0960ad52cf2a9616cb7c6a72d/bootstrap/kubeadm/main.go#L207

**Comment by jdef** [comment]
> I think it would be possible to have local env/cli-flag override for this. I can understand the user story, but it's definitely an exception.
> 
> @jdef can you use a cloud region jump box in the interim?

Nope, we're in the process of building out some VM infrastructure, but that's some O(weeks) away. And metal is scarce in our colo's right now. Easiest/fastest is to fork for now. And sync up w/ upstream once there's an option there to override this timeout.

**Comment by timothysc** [comment]
@jdef - I meant can you vpn in from say some APAC AWS region machine?  Or are your networks sealed?

**Comment by jdef** [comment]
> @jdef - I meant can you vpn in from say some APAC AWS region machine? Or are your networks sealed?

Ack. Networks are mostly air-gapped, our container images are all sourced from a private registry/repo. And I'm not on the AWS or GCP access lists yet :(  I expect that we'll be tackling related, cloudy things sometime in Q1/Q2 next year. Maybe then, problems like this will go away on their own.

**Comment by fabriziopandini** [comment]
/milestone v1.1
depending on the implementation, also candidate for cherry pick

**Comment by vincepri** [comment]
/retitle Provide ability to customize the timeout for KCP when talking to a workload etcd cluster

**Comment by neolit123** [comment]
>  It's probably reasonable for a 2s delay when connecting via localhost. Comparing use case of kubeadm-via-localhost and KCP-via-Internet seems like comparing apples and oranges.

kubeadm's etcd client also uses this 2s dial timeout to connect to remote etcd clusters.
but if there is high-latency in this scenario the apiserver will have a tough time accessing its storage backend.

for your remote management use-case configurable timeouts make sense.

**Comment by randomvariable** [comment]
I think this is reasonable. However, in the Telco space, we may want per cluster flexibility where cell sites have the higher timeout (where there's demand to be able to spin up single node clusters) but clusters within the core network use the default.

**Comment by jdef** [comment]
FWIW, in our fork we went w/ using the default of 2s unless there's an envvar defined to override.

**Comment by randomvariable** [comment]
I think if we make it considerable per cluster, we'd need a -max-ttl flag to prevent misuse.

**Comment by timoreimann** [comment]
We had the exact same need as @jdef to bump the timeout, though for slightly different reasons: the control plane of one of our CAPI-managed clusters was under a certain amount of extensive load that happened to be too large for the underlying VM to sustain. The solution we aimed for was to rolling-update the control plane to bigger instances. However, the existing load made it incredibly hard for the CAPI controllers to proxy to the etcd instance without hitting the hard-coded 2 secs limit, so the rolling update would take unreasonably long to make progress.

Forking CAPI and customizing the timeout through an environment variable is what we also went with (great minds thinking alike), though a CLI flag might be the better (because more discoverable) approach.

I could take a stab at a PR if we have consensus and nobody is on it already.
